python_modules/dagster/dagster/_core/definitions/freshness_checks/utils.py

```diff
@@ -146,16 +159,157 @@ def get_last_updated_timestamp(record: Optional[EventLogRecord]) -> Optional[flo
 
 
 def ensure_freshness_checks(checks: Sequence[AssetChecksDefinition]) -> None:
-    from .non_partitioned import NON_PARTITIONED_FRESHNESS_PARAMS_METADATA_KEY
-    from .time_window_partitioned import TIME_WINDOW_PARTITIONED_FRESHNESS_PARAMS_METADATA_KEY
-
-    for asset_check in checks:
-        for check_spec in asset_check.check_specs:
+    for check_def in checks:
+        for check_spec in check_def.check_specs:
             check.invariant(
-                check_spec.metadata
-                and (
-                    NON_PARTITIONED_FRESHNESS_PARAMS_METADATA_KEY in check_spec.metadata
-                    or TIME_WINDOW_PARTITIONED_FRESHNESS_PARAMS_METADATA_KEY in check_spec.metadata
-                ),
+                check_spec.metadata and check_spec.metadata.get(FRESHNESS_PARAMS_METADATA_KEY),
                 f"Asset check {check_spec.key} didn't have expected metadata. Please ensure that the asset check is a freshness check.",
             )
+
+
+def get_last_update_time_lower_bound(
+    freshness_cron: Optional[str],
+    current_time: datetime.datetime,
+    lower_bound_delta: datetime.timedelta,
+    freshness_cron_timezone: str,
+) -> datetime.datetime:
+    """Get the lower bound of the last_update_time for the given cron schedule."""
+    if freshness_cron:
+        return (
+            check.not_none(
+                get_latest_completed_cron_tick(
+                    freshness_cron=freshness_cron,
+                    current_time=current_time,
+                    timezone=freshness_cron_timezone,
+                )
+            )
+            - lower_bound_delta
+        )
+    else:
+        return current_time - lower_bound_delta
+
+
+def get_latest_complete_partition_key(
+    freshness_cron: Optional[str],
+    current_time: datetime.datetime,
+    freshness_cron_timezone: str,
+    partitions_def: Optional[TimeWindowPartitionsDefinition],
+) -> Optional[str]:
+    """Get the latest complete partition key for the given cron schedule."""
+    if not partitions_def:
+        return None
+    latest_cron_tick_time = check.not_none(
+        get_latest_completed_cron_tick(
+            freshness_cron=freshness_cron,
+            current_time=current_time,
+            timezone=freshness_cron_timezone,
+        )
+    )
+    time_window = check.not_none(partitions_def.get_prev_partition_window(latest_cron_tick_time))
+    return partitions_def.get_partition_key_range_for_time_window(time_window).start
+
+
+def build_freshness_check(
```

Comment:
IMO this makes things largely more confusing relative to the earlier PR.  Before, I could reason through how things worked in the different cases fairly straightforwardly. Now:
- When I'm trying to understand how freshness is defined for a non-partitioned asset, I need to reason through what partitions have to do with it (functionally: nothing).
- When I'm trying to define how freshness is defined for a partitioned asset, I need to reason through what "last update time" has to do with it (functionally: nothing).

That all said, given that this is not exposed publicly, I think the stakes are low, so I'm totally fine with shipping this version.

Reply:
@sryza is "you" in this case the user or "you" in this case the dagster framework engineer?

Reply:
It's a user who peeks under the hood to understand exactly what's going on.

Reply:
and also personally literally me trying to understand what's going on in this function

Reply:
One we move the implementation of this to use `AssetSlice` I think it will become much more clear because it will not distinguish between the partitioned and unpartitioned use case at all. It will also support dynamic, static, and multi-dimensional partitioning use cases. And it will probably be less code.

Reply:
> I think it will become much more clear because it will not distinguish between the partitioned and unpartitioned use case at all. 

IMO this is exactly the problem. The partitioned and unpartitioned use cases are actually different here, as evidenced by the behaviors, docstrings, and descriptions. Complecting them is what makes it confusing.

Reply:
My above comment should have been scoped to time-partitioned assets. I agree that each partition in a static-partitioned asset should normally function the same as a non-partitioned asset.

Reply:
Got it makes sense